From 89fcfee2c408571b957b212732a7862ae1c6cb06 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Amadeusz=20S=C5=82awi=C5=84ski?=
 <amadeuszx.slawinski@intel.com>
Date: Wed, 4 Jul 2018 09:59:30 +0200
Subject: [PATCH 229/292] ASoC: Intel: Skylake: Initialize lists before so they
 are safe to use
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

If skl_probe_work() was not run driver ends up dereferencing NULL
pointer when operating on lists in skl_platform_unregister().
To fix this initialize lists in skl_create(). Also run
cancel_work_sync() before all cleanup functions, so we don't end up
unnecessarily running probe work.

Easily reproducible with:
while true; do rmmod snd_soc_skl; modprobe snd_soc_skl; done

[  132.170160] BUG: unable to handle kernel NULL pointer dereference at 0000000000000000
[  132.170168] PGD 0 P4D 0
[  132.170176] Oops: 0000 [#1] SMP PTI
[  132.170183] CPU: 2 PID: 3967 Comm: rmmod Not tainted 4.18.0-rc3-upstream #2
[  132.170186] Hardware name: Intel Corporation CannonLake Client Platform/CannonLake Y LPDDR4 RVP, BIOS CNLSFWR1.R00.X122.B01.1801151045 01/15/2018
[  132.170199] RIP: 0010:skl_platform_unregister+0x2c/0xb0 [snd_soc_skl]
[  132.170202] Code: 44 00 00 41 55 41 54 55 53 4c 8b af 98 00 00 00 49 8b 85 50 05 00 00 4d 8d a5 50 05 00 00 49 39 c4 74 6c 49 8b bd 50 05 00 00 <48> 8b 07 49 39 fc 48 8d 5f f0 48 8d 68 f0 75 05 eb 53 48 89 c5 e8
[  132.170273] RSP: 0018:ffffa9ff0131be00 EFLAGS: 00010282
[  132.170279] RAX: 0000000000000000 RBX: ffff935e6074d018 RCX: ffff935e6074d070
[  132.170283] RDX: 0000000000000000 RSI: 0000000000000202 RDI: 0000000000000000
[  132.170287] RBP: ffff935e658c5000 R08: 0000000000000000 R09: 0000000000000000
[  132.170291] R10: ffffa9ff0131beb0 R11: ffff935e63d53701 R12: ffff935e6074d568
[  132.170294] R13: ffff935e6074d018 R14: ffff935e65f20460 R15: 0000000000000000
[  132.170299] FS:  00007f0a49026700(0000) GS:ffff935e6f900000(0000) knlGS:0000000000000000
[  132.170304] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  132.170308] CR2: 0000000000000000 CR3: 000000021d53a004 CR4: 0000000000760ee0
[  132.170312] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[  132.170316] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[  132.170320] PKRU: 55555554
[  132.170322] Call Trace:
[  132.170337]  ? skl_remove+0x48/0xb0 [snd_soc_skl]
[  132.170346]  ? pci_device_remove+0x3b/0xc0
[  132.170356]  ? device_release_driver_internal+0x17f/0x250
[  132.170363]  ? driver_detach+0x39/0x70
[  132.170370]  ? bus_remove_driver+0x51/0xd0
[  132.170376]  ? pci_unregister_driver+0x26/0xa0
[  132.170384]  ? __x64_sys_delete_module+0x153/0x1f0
[  132.170391]  ? exit_to_usermode_loop+0xac/0xf0
[  132.170398]  ? do_syscall_64+0x55/0x100
[  132.170406]  ? entry_SYSCALL_64_after_hwframe+0x44/0xa9
[  132.170411] Modules linked in: snd_soc_skl(-) ax88179_178a usbnet wmi_bmof snd_soc_skl_ipc snd_soc_sst_ipc snd_soc_sst_dsp snd_hda_ext_core snd_hda_core snd_soc_acpi snd_soc_core snd_compress snd_pcm_dmaengine ac97_bus snd_pcm intel_rapl x86_pkg_temp_thermal intel_powerclamp coretemp kvm_intel kvm snd_seq_midi snd_seq_midi_event irqbypass crct10dif_pclmul crc32_pclmul ghash_clmulni_intel pcbc snd_rawmidi aesni_intel snd_seq aes_x86_64 snd_seq_device crypto_simd snd_timer cryptd glue_helper snd hid_generic input_leds serio_raw soundcore idma64 mei_me virt_dma mei intel_lpss_pci intel_lpss intel_pch_thermal wmi acpi_pad parport_pc ppdev lp parport usbhid hid i915 nvme sdhci_pci cqhci e1000e nvme_core sdhci video [last unloaded: snd_soc_skl]
[  132.170496] CR2: 0000000000000000
[  132.170501] ---[ end trace f93ae1205bf60f85 ]---
[  132.170511] RIP: 0010:skl_platform_unregister+0x2c/0xb0 [snd_soc_skl]
[  132.170513] Code: 44 00 00 41 55 41 54 55 53 4c 8b af 98 00 00 00 49 8b 85 50 05 00 00 4d 8d a5 50 05 00 00 49 39 c4 74 6c 49 8b bd 50 05 00 00 <48> 8b 07 49 39 fc 48 8d 5f f0 48 8d 68 f0 75 05 eb 53 48 89 c5 e8
[  132.170586] RSP: 0018:ffffa9ff0131be00 EFLAGS: 00010282
[  132.170591] RAX: 0000000000000000 RBX: ffff935e6074d018 RCX: ffff935e6074d070
[  132.170595] RDX: 0000000000000000 RSI: 0000000000000202 RDI: 0000000000000000
[  132.170599] RBP: ffff935e658c5000 R08: 0000000000000000 R09: 0000000000000000
[  132.170603] R10: ffffa9ff0131beb0 R11: ffff935e63d53701 R12: ffff935e6074d568
[  132.170607] R13: ffff935e6074d018 R14: ffff935e65f20460 R15: 0000000000000000
[  132.170612] FS:  00007f0a49026700(0000) GS:ffff935e6f900000(0000) knlGS:0000000000000000
[  132.170617] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  132.170621] CR2: 0000000000000000 CR3: 000000021d53a004 CR4: 0000000000760ee0
[  132.170625] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[  132.170629] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[  132.170632] PKRU: 55555554

Change-Id: Ieff74c9c3af4ef300e63fb37f8e1f62d7552f5ee
Signed-off-by: Amadeusz Sławiński <amadeuszx.slawinski@intel.com>
Reviewed-on:
Reviewed-by: Lewandowski, Gustaw <gustaw.lewandowski@intel.com>
---
 sound/soc/intel/skylake/skl-pcm.c | 3 ---
 sound/soc/intel/skylake/skl.c     | 5 ++++-
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/sound/soc/intel/skylake/skl-pcm.c b/sound/soc/intel/skylake/skl-pcm.c
index aae3d06cffd6..7c83f05d23bb 100644
--- a/sound/soc/intel/skylake/skl-pcm.c
+++ b/sound/soc/intel/skylake/skl-pcm.c
@@ -2149,9 +2149,6 @@ int skl_platform_component_register(struct device *dev,
 	int i, j, index;
 	bool dai_already_added;
 
-	INIT_LIST_HEAD(&skl->ppl_list);
-	INIT_LIST_HEAD(&skl->bind_list);
-
 	skl->grp_cnt.vbus_id = devm_kcalloc(dev, skl->nhlt->endpoint_count,
 						sizeof(int), GFP_KERNEL);
 	if (!skl->grp_cnt.vbus_id) {
diff --git a/sound/soc/intel/skylake/skl.c b/sound/soc/intel/skylake/skl.c
index 3088e93cf5b3..046d839ea78f 100644
--- a/sound/soc/intel/skylake/skl.c
+++ b/sound/soc/intel/skylake/skl.c
@@ -514,7 +514,6 @@ static int skl_free(struct hdac_bus *bus)
 
 	snd_hdac_ext_bus_exit(bus);
 
-	cancel_work_sync(&skl->probe_work);
 	if (IS_ENABLED(CONFIG_SND_SOC_HDAC_HDMI)) {
 		snd_hdac_display_power(bus, HDA_CODEC_IDX_CONTROLLER, false);
 		snd_hdac_i915_exit(bus);
@@ -945,6 +944,9 @@ static int skl_create(struct pci_dev *pci,
 	hbus = skl_to_hbus(skl);
 	bus = skl_to_bus(skl);
 
+	INIT_LIST_HEAD(&skl->ppl_list);
+	INIT_LIST_HEAD(&skl->bind_list);
+
 #if IS_ENABLED(CONFIG_SND_SOC_INTEL_SKYLAKE_HDAUDIO_CODEC)
 	ext_ops = snd_soc_hdac_hda_get_ops();
 #endif
@@ -1210,6 +1212,7 @@ static void skl_remove(struct pci_dev *pci)
 	struct hdac_bus *bus = pci_get_drvdata(pci);
 	struct skl *skl = bus_to_skl(bus);
 
+	cancel_work_sync(&skl->probe_work);
 	skl_delete_notify_kctl_list(skl->skl_sst);
 	release_firmware(skl->tplg);
 
-- 
2.21.0

